From acb30d355f9ea6f3b41dacfbdaf70eea177ea62c Mon Sep 17 00:00:00 2001
From: Marcell Haritopoulos <marcellharitopoulos@gmail.com>
Date: Fri, 31 Jan 2025 05:35:24 +0100
Subject: [PATCH 03/13] filemap: mark pages accessed by task in bitfield

---
 include/linux/pagemap.h |   7 ++
 mm/filemap.c            | 148 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 155 insertions(+)

diff --git a/include/linux/pagemap.h b/include/linux/pagemap.h
index 09b581c1d878..70d4b461621f 100644
--- a/include/linux/pagemap.h
+++ b/include/linux/pagemap.h
@@ -1268,6 +1268,13 @@ int add_to_page_cache_lru(struct page *page, struct address_space *mapping,
 		pgoff_t index, gfp_t gfp);
 int filemap_add_folio(struct address_space *mapping, struct folio *folio,
 		pgoff_t index, gfp_t gfp);
+
+#ifdef CONFIG_PAGE_CACHE_ISOLATION
+int filemap_set_accessed_bits(struct address_space *mapping,
+		struct mm_struct *mm, off_t start, off_t len);
+#endif /* CONFIG_PAGE_CACHE_ISOLATION */
+
+
 void filemap_remove_folio(struct folio *folio);
 void __filemap_remove_folio(struct folio *folio, void *shadow);
 void replace_page_cache_folio(struct folio *old, struct folio *new);
diff --git a/mm/filemap.c b/mm/filemap.c
index 024b71da5224..01d54fbc1cc3 100644
--- a/mm/filemap.c
+++ b/mm/filemap.c
@@ -10,6 +10,13 @@
  * most "normal" filesystems (but you don't /have/ to use this:
  * the NFS filesystem used to do this differently, for example)
  */
+#include "linux/debug_locks.h"
+#include "linux/err.h"
+#include "linux/gfp_types.h"
+#include "linux/maple_tree.h"
+#include "linux/printk.h"
+#include "linux/slab.h"
+#include "linux/xarray.h"
 #include <linux/export.h>
 #include <linux/compiler.h>
 #include <linux/dax.h>
@@ -125,6 +132,103 @@
  *    ->private_lock		(zap_pte_range->block_dirty_folio)
  */
 
+#ifdef CONFIG_PAGE_CACHE_ISOLATION
+static int filemap_get_access_bitmask(struct address_space *mapping,
+		struct mm_struct *mm, struct xarray **out)
+{
+	int ret;
+	unsigned long key;
+	struct xarray* bits_arr;
+
+	*out = NULL;
+
+	key = (size_t)mm & ULONG_MAX; // TODO!
+
+	// Try to acquire the xarray based on the mm
+retry:
+	bits_arr = mtree_load(&mapping->i_mm_access_bits, key);
+	if (!bits_arr) {
+		// TODO: slab cache
+		bits_arr = kzalloc(sizeof(struct xarray), GFP_KERNEL);
+		if (!bits_arr)
+			return -ENOMEM;
+
+		xa_init_flags(bits_arr, XA_FLAGS_LOCK_IRQ);
+
+		ret = mtree_insert(&mapping->i_mm_access_bits, key, bits_arr, GFP_KERNEL);
+		if (ret == -EEXIST) {
+			kfree(bits_arr);
+			goto retry;
+		}
+		if (ret) {
+			kfree(bits_arr);
+			return ret;
+		}
+	}
+
+	*out = bits_arr;
+	return 0;
+}
+
+static int _filemap_set_accessed_bits(struct xarray *bits_arr,
+		off_t start, off_t len, size_t *non_accessed_folios)
+{
+	off_t pg_bit;
+	pgoff_t pg_idx;
+	unsigned long bitfield;
+	unsigned long mask;
+	void* field_dat;
+
+	if (non_accessed_folios)
+		*non_accessed_folios = 0;
+
+	for (pg_bit = start; pg_bit < start + len; pg_bit++) {
+		pg_idx = pg_bit / sizeof(unsigned long);
+
+		xa_lock_irq(bits_arr);
+
+		field_dat = xa_load(bits_arr, pg_idx);
+		if (xa_is_value(field_dat)) {
+			bitfield = xa_to_value(field_dat);
+		} else {
+			bitfield = 0;
+		}
+
+		mask = (1 << (pg_bit % sizeof(unsigned long)));
+		if (!(bitfield & mask)) {
+			bitfield |= mask;
+			if (non_accessed_folios)
+				*non_accessed_folios += 1;
+			field_dat = __xa_store(bits_arr, pg_idx, xa_mk_value(bitfield), GFP_KERNEL);
+		}
+
+		xa_unlock_irq(bits_arr);
+
+		if (xa_err(field_dat)) {
+			return xa_err(field_dat);
+		}
+	}
+
+	return 0;
+}
+
+
+
+int filemap_set_accessed_bits(struct address_space *mapping,
+		struct mm_struct *mm, off_t start, off_t len)
+{
+	int ret;
+	struct xarray* bits_arr;
+
+	ret = filemap_get_access_bitmask(mapping, mm, &bits_arr);
+	if (ret)
+		return ret;
+
+	return _filemap_set_accessed_bits(bits_arr, start, len, NULL);
+}
+#endif /* CONFIG_PAGE_CACHE_ISOLATION */
+
+
 static void page_cache_delete(struct address_space *mapping,
 				   struct folio *folio, void *shadow)
 {
@@ -2607,6 +2711,40 @@ static int filemap_readahead(struct kiocb *iocb, struct file *file,
 	return 0;
 }
 
+
+#ifdef CONFIG_PAGE_CACHE_ISOLATION
+static int filemap_set_accessed_bits_fbatch(struct address_space *mapping,
+		struct folio_batch *fbatch, struct mm_struct *mm, 
+		size_t *non_accessed_folios)
+{
+	int i, ret;
+	size_t folio_non_accessed;
+	pgoff_t idx, cnt;
+	struct xarray *bits_arr;
+
+	*non_accessed_folios = 0;
+
+
+	ret = filemap_get_access_bitmask(mapping, mm, &bits_arr);
+	if (ret)
+		return ret;
+
+	for (i = 0; i < folio_batch_count(fbatch); i++) {
+		idx = folio_index(fbatch->folios[i]);
+		cnt = folio_nr_pages(fbatch->folios[i]);
+
+		ret = _filemap_set_accessed_bits(bits_arr, idx, cnt, &folio_non_accessed);
+		if (ret)
+			return ret;
+
+		*non_accessed_folios += folio_non_accessed;
+	}
+
+
+	return 0;
+}
+#endif
+
 static int filemap_get_pages(struct kiocb *iocb, size_t count,
 		struct folio_batch *fbatch, bool need_uptodate)
 {
@@ -2616,6 +2754,9 @@ static int filemap_get_pages(struct kiocb *iocb, size_t count,
 	pgoff_t last_index;
 	struct folio *folio;
 	unsigned int flags;
+#ifdef CONFIG_PAGE_CACHE_ISOLATION
+	size_t non_accessed_folios = 0;
+#endif /* CONFIG_PAGE_CACHE_ISOLATION */
 	int err = 0;
 
 	/* "last_index" is the index of the folio beyond the end of the read */
@@ -2664,6 +2805,13 @@ static int filemap_get_pages(struct kiocb *iocb, size_t count,
 			goto err;
 	}
 
+#ifdef CONFIG_PAGE_CACHE_ISOLATION
+	/* Mark the folios as accessed to the  */
+	err = filemap_set_accessed_bits_fbatch(mapping, fbatch, current->mm, &non_accessed_folios);
+	if (err)
+		goto err;
+#endif /* CONFIG_PAGE_CACHE_ISOLATION */
+
 	trace_mm_filemap_get_pages(mapping, index, last_index - 1);
 	return 0;
 err:
-- 
2.55.0

